File: code/go/bench.go
package main

import (
	"fmt"
	"math"
	rnd "math/rand"
	"time"
)

func sigmoid(arg float64) float64 {
	return 1. / (1. + math.Exp(-arg))
}

func tanh(arg float64) float64 {
	a1 := math.Exp(arg)
	a2 := math.Exp(-arg)
	return (a1 - a2) / (a1 + a2)
}

func main() {
	rnd.Seed(2019)
	length := 1000000
	random_max := 100
	var input = make([]int, length)
	var output = make([]float64, length)

	for i := 0; i < length; i++ {
		input[i] = rnd.Intn(random_max)
	}

	// multiplication
	t0 := time.Now()
	for i := 0; i < length; i++ {
		output[i] = float64(input[i] * 2.)
	}
	t := time.Since(t0)
	fmt.Println("Multiplication by 2:", t)

	// sum
	t0 = time.Now()
	for i := 0; i < length; i++ {
		if i == length-1 {
			output[i] = float64(input[i] + input[0])
		} else {
			output[i] = float64(input[i] + input[i+1])
		}
	}
	t = time.Since(t0)
	fmt.Println("Sum:", t)

	// power
	t0 = time.Now()
	for i := 0; i < length; i++ {
		output[i] = math.Pow(float64(input[i]), 2.)
	}
	t = time.Since(t0)
	fmt.Println("Power of 2:", t)

	// sigmoid
	t0 = time.Now()
	for i := 0; i < length; i++ {
		output[i] = sigmoid(float64(input[i]))
	}
	t = time.Since(t0)
	fmt.Println("Sigmoid:", t)

	// sigmoid
	t0 = time.Now()
	for i := 0; i < length; i++ {
		output[i] = tanh(float64(input[i]))
	}
	t = time.Since(t0)
	fmt.Println("Tanh:", t)
}

File: code/r/bench.R
sigmoid <- function(arg) {
  1 / (1 + exp(arg))
}

tanh <- function(arg) {
  arg1 <- exp(arg)
  arg2 <- exp(-arg)

  return ( (arg1 - arg2)/(arg1 + arg2) )
}

set.seed(2019)
length <- 1000000
random_max <- 100

input <- sample(0:random_max, length, replace = TRUE)

# multiplication
t0 <- Sys.time()
output <- input * 2.
t <- (Sys.time() - t0) * 1e3
cat(paste0("Multiplication by 2: ", t, "ms\n"))

# sum
t0 <- Sys.time()
output <- input + input[c(2:length, 1)]
t <- (Sys.time() - t0) * 1e3
cat(paste0("Sum: ", t, "ms\n"))

# power
t0 <- Sys.time()
output <- input^2
t <- (Sys.time() - t0) * 1e3
cat(paste0("Power of 2: ", t, "ms\n"))

# sigmoid
t0 <- Sys.time()
output <- sigmoid(input)
t <- (Sys.time() - t0) * 1e3
cat(paste0("Sigmoid: ", t, "ms\n"))

# tanh
t0 <- Sys.time()
output <- tanh(input)
t <- (Sys.time() - t0) * 1e3
cat(paste0("Tanh: ", t, "ms\n"))
